### PR TITLE
Feature/69 clean up

### LINKED ISSUE #69: Refactor ManagementClientTests Architecture
### Motivation

* Currently, the tests are hard to maintain, read, and basically dependent on the live project in Kontent if new features are added and bugs are fixed.
* Re-generation requires there steps - Re-run ManagementClientTests with `private static readonly TestUtils.TestRunType _runType = TestUtils.TestRunType.MockFromFileSystem;`, after tests are green, commit changes in `~/Kentico.Kontent.Management.Tests/Data/*`

Come up with a better approach that would test the SDK functionality without necessity to pre-record requests and reposnse.

### Proposed solution

These are just possible enhancements:
* [x] Do now generate hash under `Data` folder - identify mock responses just by test name - so that update does not generate folder removal and addiction - and this pair is hard to distinguish
* [ ] Automate regeneration of the mock files
* [ ] Hardcode the JSONs and avoid auto-generation completely

### Additional context

* We might use [MAPI Swagger definition](https://kcdmaster.blob.core.windows.net/api-reference-specifications/management_api_v2.json) and use it for mock server -  (if URL does not work - reach out to @Simply007)

### Related issues

#79 #66 #64

### PR DESCRIPTION
**Pull request by gormal**
### Motivation

Fixes #69 

This is an overall clean-up of the solution as we are getting ready for a major release.

### CODE REVIEW

File: Kentico.Kontent.Management.Tests/Base/FileSystemFixture.cs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+namespace Kentico.Kontent.Management.Tests.Base;
+
+public sealed class FileSystemFixture : IDisposable
+{
+    public static string Endpoint => "https://manage.kontent.ai/v2";
+    public static string PROJECT_ID => "a9931a80-9af4-010b-0590-ecb1273cf1b8";
+    public static string SUBCRIPTION_ID => "9c7b9841-ea99-48a7-a46d-65b2549d6c0";
+
+    private string _folder = "";
+    private readonly ManagementOptions _managementOptions;
```
Comment (winklertomas) [inline Kentico.Kontent.Management.Tests/Base/FileSystemFixture.cs:25]: can be converted to local variable

File: Kentico.Kontent.Management.Tests/ManagementHttpClientTests.cs [hunk 2]
```diff
@@ -11,168 +11,166 @@
 using System.Threading.Tasks;
 using Xunit;
```
Comment (winklertomas) [inline Kentico.Kontent.Management.Tests/ManagementHttpClientTests.cs:13]: unused using System.Collections.Generic;

**Review by winklertomas — CHANGES_REQUESTED** [review]

File: Kentico.Kontent.Management.Tests/CodeSamples/Readme.cs [hunk 3]
```diff
@@ -69,7 +69,7 @@ public Readme(FileSystemFixture fileSystemFixture)
         public void CreateManagementClient()
         {
             // Initializes an instance of the ManagementClient client with specified options.
-            _ = new ManagementClient(new ManagementOptions
```
[the remark this answers is not in the document]
Comment (gormal) [reply Kentico.Kontent.Management.Tests/CodeSamples/Readme.cs]: This is file should contain code samples used in Readme.md and it just looks better with `var client`

File: Kentico.Kontent.Management.Tests/ManagementClientTests/AssetTests.cs [hunk 4]
```diff
@@ -0,0 +1,684 @@
...
+    [Fact]
+    public async Task ListAssetsAsync_DynamicallyTyped_WithMorePages_ListsAssets()
+    {
+        var client = _fileSystemFixture.CreateMockClientWithResponse("AssetsPage1.json", "AssetsPage2.json", "AssetsPage3.json");
+
+        var expected = new[] {
+            "00000000-0000-0000-0000-000000000000",
+            "10000000-0000-0000-0000-000000000000",
+            "20000000-0000-0000-0000-000000000000"
+        }.Select(x => GetExpectedDynamicAssetModel(assetId: x));
```
[the remark this answers is not in the document]
Comment (gormal) [reply Kentico.Kontent.Management.Tests/ManagementClientTests/AssetTests.cs:36]: The build-in analyzer does not offer this refactoring, so I won't enforce it.

File: Kentico.Kontent.Management.Tests/ManagementClientTests/AssetTests.cs [hunk 5]
```diff
@@ -0,0 +1,684 @@
...
+    [Fact]
+    public async Task ListAssetsAsync_StronglyTyped_WithMorePages_ListsAssets()
+    {
+        var client = _fileSystemFixture.CreateMockClientWithResponse("AssetsPage1.json", "AssetsPage2.json", "AssetsPage3.json");
+
+        var expected = new[] {
+            "00000000-0000-0000-0000-000000000000",
+            "10000000-0000-0000-0000-000000000000",
+            "20000000-0000-0000-0000-000000000000"
+        }.Select(x => GetExpectedStronglyTypedAssetModel(assetId: x));
```
[the remark this answers is not in the document]
Comment (gormal) [reply Kentico.Kontent.Management.Tests/ManagementClientTests/AssetTests.cs:52]: The build-in analyzer does not offer this refactoring, so I won't enforce it.

File: Kentico.Kontent.Management/ManagementClient.cs [hunk 6]
```diff
@@ -27,1212 +27,1212 @@
...
+        var response = await _actionInvoker.InvokeReadOnlyMethodAsync<LanguageVariantsListingResponseServerModel>(endpointUrl, HttpMethod.Get);
 
-            return new ListingResponseModel<LanguageVariantModel>(
-                    (token, url) => GetNextListingPageAsync<LanguageVariantsListingResponseServerModel, LanguageVariantModel>(token, url),
-                    response.Pagination?.Token,
-                    endpointUrl,
-                    response.Variants);
-        }
+        return new ListingResponseModel<LanguageVariantModel>(
+                (token, url) => GetNextListingPageAsync<LanguageVariantsListingResponseServerModel, LanguageVariantModel>(token, url),
```
[the remark this answers is not in the document]
Comment (gormal) [reply Kentico.Kontent.Management/ManagementClient.cs:333]: The build-in analyzer does not offer this refactoring, so I won't enforce it.

File: Kentico.Kontent.Management/Modules/ActionInvoker/DynamicObjectJsonConverter.cs [hunk 7]
```diff
@@ -5,73 +5,61 @@
...
-                    continue;
-                }
-
-                var value = property.Value as JValue;
-                if (value != null)
-                {
-                    resultAsDictionary.Add(property.Name, value.Value);
-                    continue;
-                }
+                resultAsDictionary.Add(property.Name, array.Select(arrayObject => ConvertJComplexObject(arrayObject)));
```
[the remark this answers is not in the document]
Comment (gormal) [reply Kentico.Kontent.Management/Modules/ActionInvoker/DynamicObjectJsonConverter.cs:24]: The build-in analyzer does not offer this refactoring, so I won't enforce it.

File: Kentico.Kontent.Management/Modules/ActionInvoker/DynamicObjectJsonConverter.cs [hunk 8]
```diff
@@ -5,73 +5,61 @@
...
-            }
-        }
 
-        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
-        {
-            var defaultResult = serializer.Deserialize(reader);
+    private static object ConvertJComplexObject(object input) => input switch
+    {
+        JObject obj => ConvertToDynamicObject(obj),
+        JArray array => array.Select(obj => ConvertJComplexObject(obj)),
```
[the remark this answers is not in the document]
Comment (gormal) [reply Kentico.Kontent.Management/Modules/ActionInvoker/DynamicObjectJsonConverter.cs:51]: The build-in analyzer does not offer this refactoring, so I won't enforce it.

File: Kentico.Kontent.Management.Tests/ManagementClientTests/ContentTypeTests.cs [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+
+        response.Should().BeEquivalentTo(expected);
+    }
+
+    [Fact]
+    public async void ModifyContentTypeAsync_IdentifierIsNull_Throws()
+    {
+        var client = _fileSystemFixture.CreateMockClientWithoutResponse();
+
+        List<ContentTypeOperationBaseModel> changes = new() { };
```
[the remark this answers is not in the document]
Comment (gormal) [reply Kentico.Kontent.Management.Tests/ManagementClientTests/ContentTypeTests.cs:143]: fixed

File: Kentico.Kontent.Management/Models/Subscription/SubscriptionProjectListingResponseServerModel.cs [hunk 10]
```diff
@@ -3,25 +3,18 @@
...
-        [JsonProperty("pagination")]
-        public PaginationResponseModel Pagination { get; set; }
+    [JsonProperty("pagination")]
+    public PaginationResponseModel Pagination { get; set; }
 
-        public IEnumerator<SubscriptionProjectModel> GetEnumerator()
-        {
-            return GetEnumerator();
-        }
+    public IEnumerator<SubscriptionProjectModel> GetEnumerator() => GetEnumerator();
```
[the remark this answers is not in the document]
Comment (gormal) [reply Kentico.Kontent.Management/Models/Subscription/SubscriptionProjectListingResponseServerModel.cs:17]: Uff, good catch :)

File: Kentico.Kontent.Management.Tests/Base/FileSystemFixture.cs (hunk 1 above)
Comment (gormal) [reply Kentico.Kontent.Management.Tests/Base/FileSystemFixture.cs:25]: fixed

File: Kentico.Kontent.Management.Tests/CodeSamples/CmApiV2.cs [hunk 11]
```diff
@@ -0,0 +1,1792 @@
...
+        var client = _fileSystemFixture.CreateMockClientWithResponse("PostTaxonomyGroupResponse.json");
+
+        var response = await client.CreateTaxonomyGroupAsync(new TaxonomyGroupCreateModel
+        {
+            Name = "Personas",
+            ExternalId = "Tax-Group-123",
+            Codename = "people",
+            Terms = new TaxonomyTermCreateModel[]
+                {
+                    new TaxonomyTermCreateModel
```
[the remark this answers is not in the document]
Comment (gormal) [reply Kentico.Kontent.Management.Tests/CodeSamples/CmApiV2.cs:1097]: The build-in analyzer didn't pick this up, so I will leave it as I want to be independent of R#

File: Kentico.Kontent.Management.Helpers.Tests/EditLinkBuilderTests.cs [hunk 12]
```diff
@@ -5,148 +5,147 @@
...
+    [Fact]
+    public void BuildEditItemUrl_MultipleElements_ReturnsValidUrl()
+    {
+        var elements = new ElementIdentifier[]
         {
-            var elementIdentifier = new ElementIdentifier(_itemId, "single-element-Codename");
-            
-            var underTest = new EditLinkBuilder(DefaultOptions);
-            var result = underTest.BuildEditItemUrl(_language, elementIdentifier);
+            new ElementIdentifier("76c06b74-bae9-4732-b629-1a59395e893d", "some-Element-Codename-1"),
```
[the remark this answers is not in the document]
Comment (gormal) [reply Kentico.Kontent.Management.Helpers.Tests/EditLinkBuilderTests.cs:111]: The build-in analyzer didn't pick this up, so I will leave it as I want to be independent of R#

File: Kentico.Kontent.Management/Modules/ActionInvoker/DynamicObjectJsonConverter.cs [hunk 13]
```diff
@@ -5,73 +5,61 @@
...
+        _ => throw new ArgumentOutOfRangeException("JSON deserialization did not return either object nor array."),
```
[the remark this answers is not in the document]
Comment (gormal) [reply Kentico.Kontent.Management/Modules/ActionInvoker/DynamicObjectJsonConverter.cs:52]: fixed

File: Kentico.Kontent.Management/Modules/ModelBuilders/ElementBuilder.cs [hunk 14]
```diff
@@ -4,37 +4,31 @@
...
             {
-                if (elementObject.Element == null)
-                {
-                    elementExceptions.Add(new ArgumentNullException("Element identifier (`BaseElement.Element` property) not set for element on index ", Array.IndexOf(elements, elementObject).ToString()));
-                }
-                else if (!elementObject.Element.DoesHaveSetOnlyOneIdentifier())
-                {
-                    elementExceptions.Add(new ArgumentException("Element must have only one identifier set (`BaseElement.Element` property).", nameof(elementObject)));
-                }
+                elementExceptions.Add(new ArgumentNullException("Element identifier (`BaseElement.Element` property) not set for element on index ", Array.IndexOf(elements, elementObject).ToString()));
```
[the remark this answers is not in the document]
Comment (gormal) [reply Kentico.Kontent.Management/Modules/ModelBuilders/ElementBuilder.cs:24]: fixed

File: Kentico.Kontent.Management.Tests/Data/ElementsData.cs [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+
+    public static ComplexTestModel GetExpectedStronglyTypedElementsModel() => new()
+    {
+        BodyCopy = new RichTextElement
+        {
+            Element = Reference.ById(typeof(ComplexTestModel).GetProperty(nameof(ComplexTestModel.BodyCopy)).GetKontentElementId()),
+            Value = $"<h1>Light Roasts</h1><p>Usually roasted for 6 - 8 minutes or simply until achieving a light brown color.This method is used for milder coffee varieties and for coffee tasting.This type of roasting allows the natural characteristics of each coffee to show.The aroma of coffees produced from light roasts is usually more intense.The cup itself is more acidic and the concentration of caffeine is higher.</p><object type=\"application/kenticocloud\" data-type=\"component\" data-id=\"{Guid.Parse("46c05bd9-d418-4507-836c-9accc5a39db3")}\"></object>",
+            Components = new ComponentModel[]
+        {
+            new ComponentModel
```
[the remark this answers is not in the document]
Comment (winklertomas) [reply Kentico.Kontent.Management.Tests/Data/ElementsData.cs]: Yes the R# suggests that.

File: Kentico.Kontent.Management.Tests/ManagementClientTests/ContentTypeSnippetTests.cs [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+using Kentico.Kontent.Management.Models.Types.Elements;
+using Kentico.Kontent.Management.Models.TypeSnippets;
+using Kentico.Kontent.Management.Models.TypeSnippets.Patch;
+using Kentico.Kontent.Management.Tests.Base;
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace Kentico.Kontent.Management.Tests.Unit.ManagementClientTests;
```
[the remark this answers is not in the document]
Comment (gormal) [reply Kentico.Kontent.Management.Tests/ManagementClientTests/ContentTypeSnippetTests.cs]: fixed

**Review by winklertomas — APPROVED** [review]
Review ok, feel free to merge it :-)

**Comment by gormal** [comment]
This can be merged only after #163.

**Pull request merged**